src/line_parser2.rs

```diff
@@ -0,0 +1,390 @@
+use crate::{BeatNumber, Line, LineNote, Result};
+use combine::{
+    choice, many, many1, optional,
+    parser::char::{digit, space, spaces},
+    sep_by, token, Parser,
+};
+
+#[derive(Clone, Copy, Debug)]
+enum Letter {
+    A,
+    B,
+    C,
+    D,
+    E,
+    F,
+    G,
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Modifier {
+    Flat,
+    Natural,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct Note {
+    letter: Letter,
+    modifier: Modifier,
+    octave: i8,
+    duration: u32,
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Value {
+    Note(Note),
+    Rest,
+}
+
+impl Note {
+    fn to_wmidi_note(self) -> wmidi::Note {
+        let letter_value = match self.letter {
+            Letter::C => 0,
+            Letter::D => 2,
+            Letter::E => 4,
+            Letter::F => 5,
+            Letter::G => 7,
+            Letter::A => 9,
+            Letter::B => 11,
+        };
+
+        let modifier_value = match self.modifier {
+            Modifier::Flat => -1,
+            Modifier::Natural => 0,
+        };
+
+        let octave_value = self.octave + 1;
+
+        let value = octave_value * 12 + letter_value + modifier_value;
+        wmidi::Note::from_u8_lossy(value as u8)
+    }
+}
+
+fn parse_line(line: &str) -> Result<Line> {
+    let letter_parser = choice((
```

Comment:
So `choice()` is accepting an arbitrary-length tuple? How does that work on a type level?

Reply:
I think they've probably written (or hopefully code-generated) implementations of their parser trait for a 2-element tuple containing 2 parser trait values, 3-element tuple, 4-element tuple, ...